File: test_succession.js
const { calculatePlantsNeeded } = require('./src/services/homeGardenCalculator.js');
const cropsRaw = require('./src/data/crops.json').crops;

function verifyTPSuccession() {
    // Force date to March 1, 2026
    const RealDate = Date;
    global.Date = class extends RealDate {
        constructor(...args) {
            if (args.length) super(...args);
            else super('2026-03-01T12:00:00Z');
        }
    };
    
    // Pick Buttercrunch Lettuce (TP crop)
    const buttercrunch = cropsRaw.find(c => c.id === 'lettuce_buttercrunch');
    
    // A standard garden profile with 150 frost-free days
    const gardenProfile = {
        last_frost_date: '2026-04-15',   // Mid-April
        first_frost_date: '2026-10-15',  // Mid-October
        frost_free_days: 180
    };
    
    console.log("--- Testing Transplant Succession Logic ---");
    console.log(`Crop: ${buttercrunch.name}`);
    console.log(`Seed Type: ${buttercrunch.seed_type}`);
    console.log(`Interval: ${buttercrunch.harvest_window_days} days`);
    console.log(`Weeks Indoors: ${buttercrunch.seed_start_weeks_before_transplant} weeks`);
    console.log(`DTM (from transplant): ${buttercrunch.dtm} days\n`);

    const result = calculatePlantsNeeded(buttercrunch, 4, gardenProfile);
    
    console.log(`Round 1 (Ideal Indoor Seed Date): ${result.indoorSeedDate}`);
    if (result.successionDates && result.successionDates.length > 0) {
        console.log("Succession Rounds Built Successfully:");
        result.successionDates.forEach(r => {
            console.log(`  Round ${r.round}: ${r.dateRaw}`);
        });
        console.log("\n✅ Parity Achieved! TP crops now model succession arrays natively.");
    } else {
        console.log("❌ FAILED. No explicit rounds calculated.");
        process.exit(1);
    }
}

try {
    verifyTPSuccession();
} catch (e) {
    console.error(e);
    process.exit(1);
}

File: test_succession.py
import json
from datetime import datetime, timedelta

with open('src/data/crops.json') as f:
    crops = json.load(f).get('crops', [])

buttercrunch = next((c for c in crops if c['id'] == 'lettuce_buttercrunch'), None)

# Fake input data mimicking the logic
seed_type = buttercrunch.get('seed_type')  # 'TP'
dtm = buttercrunch.get('dtm')              # 65
tp_weeks = buttercrunch.get('seed_start_weeks_before_transplant') # 4
interval = buttercrunch.get('harvest_window_days', 14) # 14

indoor_seed_date_raw = datetime(2026, 3, 14) 
first_frost = datetime(2026, 10, 15)

print("--- Testing Transplant Succession Logic ---")
print(f"Crop: {buttercrunch['name']}")
print(f"Seed Type: {seed_type}")
print(f"Interval: {interval} days")
print(f"Weeks Indoors: {tp_weeks} weeks")
print(f"DTM (from transplant): {dtm} days\n")

succession_dates = []

# Exact mimic of homeGardenCalculator.js patched loop
anchor_sow_raw = indoor_seed_date_raw if seed_type == 'TP' else None

if anchor_sow_raw and first_frost:
    for round_num in range(2, 9):
        sow_raw = anchor_sow_raw + timedelta(days=(round_num - 1) * interval)
        
        if seed_type == 'TP':
            # Maturity from seed = seed_start_weeks + dtm
            maturity_raw = sow_raw + timedelta(days=(tp_weeks * 7) + dtm)
        else:
            maturity_raw = sow_raw + timedelta(days=dtm)
            
        if maturity_raw > first_frost:
            break
            
        succession_dates.append({
            'round': round_num,
            'dateRaw': sow_raw.strftime('%Y-%m-%d'),
            'maturity_theoretical': maturity_raw.strftime('%Y-%m-%d')
        })

print(f"Round 1 (Ideal Indoor Seed Date): {indoor_seed_date_raw.strftime('%Y-%m-%d')}")
if succession_dates:
    print("Succession Rounds Built Successfully:")
    for r in succession_dates:
        print(f"  Round {r['round']}: Seed Indoors {r['dateRaw']} -> Matures before frost: {r['maturity_theoretical']}")
    print("\n✅ Parity Achieved! TP crops now model exact succession arrays natively.")
else:
    print("❌ FAILED. No explicit rounds calculated.")
